include/GraphUpscale.hpp

```diff
@@ -0,0 +1,318 @@
+/*BHEADER**********************************************************************
+ *
+ * Copyright (c) 2018, Lawrence Livermore National Security, LLC.
+ * Produced at the Lawrence Livermore National Laboratory.
+ * LLNL-CODE-745247. All Rights reserved. See file COPYRIGHT for details.
+ *
+ * This file is part of smoothG. For more information and source code
+ * availability, see https://www.github.com/llnl/smoothG.
+ *
+ * smoothG is free software; you can redistribute it and/or modify it under the
+ * terms of the GNU Lesser General Public License (as published by the Free
+ * Software Foundation) version 2.1 dated February 1999.
+ *
+ ***********************************************************************EHEADER*/
+
+/** @file
+
+    @brief Contains GraphUpscale class
+*/
+
+#ifndef __GRAPHUPSCALE_HPP__
+#define __GRAPHUPSCALE_HPP__
+
+#include "linalgcpp.hpp"
+#include "parlinalgcpp.hpp"
+#include "partition.hpp"
+
+#include "Utilities.hpp"
+#include "MixedMatrix.hpp"
+#include "GraphCoarsen.hpp"
+#include "MGLSolver.hpp"
+#include "Graph.hpp"
+
+#include "MinresBlockSolver.hpp"
+#include "HybridSolver.hpp"
+#include "SPDSolver.hpp"
+
+namespace smoothg
+{
+
+/**
+   @brief Use upscaling as operator
+*/
+
+class GraphUpscale : public linalgcpp::Operator
+{
+public:
+    /// Default Constructor
+    GraphUpscale() = default;
+
+    /**
+       @brief Graph Constructor
+
+       @param graph contains input graph information
+       @param spect_tol spectral tolerance determines how many eigenvectors to
+                        keep per aggregate
+       @param max_evects maximum number of eigenvectors to keep per aggregate
+       @param hybridization use hybridization as solver
+    */
+    GraphUpscale(Graph graph, double spect_tol = 0.001, int max_evects = 4,
+                 bool hybridization = false, const std::vector<int>& elim_edge_dofs = {});
```

Comment:
Currently there is no example that uses `elim_edge_dof`, right? Since `GraphTopology` does not produce "coarse boundary", we need to be careful. If I understand correctly, the boundary is identified by partition. This can produce some unexpected results if the user is not aware of the assumption. 

I guess we need to think about how to handle different essential dofs (there will be cases where vertex dofs is fixed). But let's do this in a separate PR. For this PR, am I right that `elim_edge_dofs` is always empty?

Reply:
Correct, so far the included examples don't use this.